solution-segmentation.py

```diff
@@ -0,0 +1,329 @@
+        """Constructs the model by training it on a set of feature samples
+
+        :param samples: A set of feature samples
+        :param n_components: The number of components in the mixture.
+        :param covariance_type: Type of covariance representation, one of 'spherical', 'tied', 'diag' or 'full'.
+        """
+
+        self._gmm = GaussianMixture(n_components=n_components, covariance_type=covariance_type, init_params='random')
+        self._perform_training(samples)
+
+    def _perform_training(self, samples):
+        """Trains the model"""
+        self._gmm.fit(samples)
+
+        # Compute minimum likelihood for computing distances similar to Mahalanobis distances.
+        num_dims = samples.shape[1]
+        num_comps = self._gmm.n_components
+
+        cov_type = self._gmm.covariance_type
+        if cov_type == 'spherical':
+            covariances = np.einsum('i,jk->ijk', self._gmm.covariances_, np.identity(num_dims))
+        elif cov_type == 'tied':
+            covariances = np.repeat(self._gmm.covariances_[np.newaxis, :, :], num_comps, axis=0)
+        elif cov_type == 'diag':
+            covariances = np.einsum('ij, jk->ijk', self._gmm.covariances_, np.identity(num_dims))
+        elif cov_type == 'full':
+            covariances = self._gmm.covariances_
+        else:
+            raise Exception("Unsupported covariance type")
+
+        self._c = 0
+        for mean, covar, w in zip(self._gmm.means_, covariances, self._gmm.weights_):
+            self._c += w / np.sqrt(np.linalg.det(2 * np.pi * covar))
+        self._c = np.log(self._c)
+
+    def compute_mahalanobis_distances(self, image):
+        """Computes the Mahalanobis distances for a feature image given this model"""
+
+        # At least, compute something similar to Mahalanobis distances for this model type.
+        samples = image.reshape(-1, 3)
+        mahalanobis = np.sqrt(2 * (self._c - self._gmm.score_samples(samples)))
+
+        return mahalanobis.reshape(image.shape[:2])
+
+
+def run_segmentation_lab():
+    # Set parameters.
+    use_otsu = False
+    use_adaptive_model = False
+    max_std_dev = 20
+    thresh_val = 8
+    adaptive_update_ratio = 0.1
+    model_type = MultivariateNormalModel  # or you can try GaussianMixtureModel
+
+    # Create windows and gui components.
+    cv2.namedWindow('frame', cv2.WINDOW_NORMAL)
+    cv2.namedWindow('mahalanobis_img', cv2.WINDOW_NORMAL)
+    cv2.namedWindow('segmented', cv2.WINDOW_NORMAL)
+
+    def onchange(val):
+        """Callback for trackbar"""
+        nonlocal thresh_val
+        thresh_val = val
+
+    cv2.createTrackbar('Threshold', 'frame', thresh_val, max_std_dev, onchange)
+
+    # Connect to the camera.
+    # Change to video file if you want to use that instead.
+    device_id = 0
+    cap = cv2.VideoCapture(device_id)
+    if not cap.isOpened():
+        print(f"Could not open camera {device_id}")
+        return
+    else:
+        print(f"Successfully opened camera {device_id}")
```

Comment:
Siden du hinter til at man kan bruke videofil, hva med
```suggestion
    video_source = 0
    cap = cv2.VideoCapture(video_source)
    if not cap.isOpened():
        print(f"Could not open video source {video_source}")
        return
    else:
        print(f"Successfully opened video source {video_source}")
```